app/services/katello/pulp3/repository/yum.rb

```diff
@@ -353,13 +358,21 @@ def copy_content_for_source(source_repository, options = {}) # rubocop:disable M
           blacklist_ids += modular_packages(source_repository, exclusion_modular_filters) unless exclusion_modular_filters.empty?
           content_unit_hrefs = whitelist_ids - blacklist_ids
           if content_unit_hrefs.any?
-            content_unit_hrefs += additional_content_hrefs(source_repository, content_unit_hrefs)
+            content_unit_hrefs += additional_content_hrefs(source_repository, content_unit_hrefs, true)
           end
           content_unit_hrefs += source_repository.srpms.pluck(:pulp_id)
           dependency_solving = options[:solve_dependencies] || false
           copy_units(source_repository, content_unit_hrefs.uniq, dependency_solving)
         end
 
+        def errata_to_include_from_map(repo_id_map)
```

Comment:
I was staring at this query for a while because I like to see how we can break ones like it (multiple, hand-written joins) into a more railsy approach which may also lend us some nice utility later. I came up with this modification to errata.rb

```
@@ -16,6 +16,7 @@ module Katello
     has_many :bugzillas, :class_name => "Katello::ErratumBugzilla", :dependent => :destroy, :inverse_of => :erratum
     has_many :cves, :class_name => "Katello::ErratumCve", :dependent => :destroy, :inverse_of => :erratum
     has_many :packages, :class_name => "Katello::ErratumPackage", :dependent => :destroy, :inverse_of => :erratum
+    has_many :repository_errata, :class_name => "Katello::RepositoryErratum", :dependent => :destroy, :inverse_of => :erratum
 
     scoped_search :on => :errata_id, :only_explicit => true
     scoped_search :on => :errata_id, :rename => :id, :complete_value => true, :only_explicit => true
@@ -50,6 +51,7 @@ module Katello
     scope :enhancement, -> { of_type(Erratum::ENHANCEMENT) }
     scope :modular, -> { where(:id => joins(:packages => :module_stream_errata_packages)) }
     scope :non_modular, -> { where.not(:id => modular) }
+    scope :with_errata_by_repo_ids, ->(repo_ids) { joins(:repository_errata).merge(Katello::RepositoryErratum.where(repository_id: repo_ids)) }

```

For someone like me this is way easier to understand! Now you can re-implement this method to use the new scope:

```
all_errata = Katello::Erratum.with_errata_by_repo_ids(repo_id_map.keys.flatten.uniq)
all_content_units  = ...
```

Let me know what you think - seems it generates an identical query to what you came up with

Reply:
I don't think the scope name is very good btw- maybe `in_repositories` is better?

Reply:
I think I like `in_repositories`, I'll give it a good stare and see if I come up with anything else. Going to try out the joins, it looks much cleaner to me too.

Reply:
Turns out `repository_errata` is defined through the PulpDatabaseUnit concern. so you really just need the new scope!

Reply:
May be worth adding a .distinct to the query too?